### PR TITLE
Bug/assorted bug fixes 0

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

2 minor bug fixes. see below

Fixes: #368 

Changes proposed:
- modify logic of balance function
- matmul now squeezes the output if it is the case of mv or vm

## Type of change

Please delete options that are not relevant.

- [X ] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: heat/core/linalg.py [hunk 1]
```diff
@@ -103,33 +103,64 @@ def matmul(a, b):
...
             b.comm.Allreduce(MPI.IN_PLACE, c, MPI.SUM)
-            return factories.array(c, split=b.split if a.gshape[-2] > 1 else 1)
+            c = c if not vector_flag else c.squeeze()
+            c = factories.array(c, split=b.split if a.gshape[-2] > 1 else 0)
+            return c
+
+        elif a.split == 0 and b.split is None:  # this case and the one below will only be reaching if one of them is a vector
+            c = torch.zeros((a.gshape[-2], b.lshape[1]), dtype=c_type.torch_type())
+            a_idx = a.comm.chunk(a.shape, a.split)[2]
+            c[a_idx[0]] += a._DNDarray__array @ b._DNDarray__array  #[a_idx[1].start:a_idx[1].start + a.lshape[-1], :]
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: comment here needs to be removed

File: heat/core/dndarray.py [hunk 2]
```diff
@@ -661,30 +661,34 @@ def balance_(self):
...
+            if not b:  # if the process is not balanced
+                send_amt = abs((chunk_map[pr, self.split] - lshape_map[pr, self.split]).item())
+                send_amt = send_amt if send_amt < lshape_map[pr, self.split] else lshape_map[pr, self.split]
+                if send_amt:
+                    if self.comm.rank == pr:  # send data to the next process
                         send_slice[self.split] = slice(self.lshape[self.split] - send_amt, self.lshape[self.split])
                         keep_slice[self.split] = slice(0, self.lshape[self.split] - send_amt)
 
-                        self.comm.Isend(self.__array[send_slice].clone(), dest=pr, tag=pr + self.comm.size + spr)
+                        self.comm.Send(self.__array[send_slice].clone(), dest=pr + 1, tag=pr + self.comm.size + pr + 1)
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/dndarray.py]: this isnt always the case that it _needs_ to be sent to the next rank but it is a save way to do it. it will cause a higher shuffle but this is safe (unless the data to be split between processes n+1 -> size -1 is much larger than is available on processes n+1)

[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/core/dndarray.py]: Right, but wouldn't we want to do this in parallel using an Alltoallv instead of an entirely sequential push-through?

[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/dndarray.py]: my goal was to avoid using Alltoallv to avoid the cost that it carries. I have the impression that in most cases i feel that it would be much heavier than is frequently required. if the sliced array is mostly balanced then an alltoall would be a superfluous call, however if thearray is very imbalanced than an alltoall makes more sense. what do you think? do you think that the balance function should just always call alltoall or is it better to have the logic hardcoded?

[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/core/dndarray.py]: Fair point. We should keep it then as is and potentially recode it using alltoallv later if necessary
